File: docquad-integration/src/docquad/DocQuadOrtRunner.java
package docquad;

import ai.onnxruntime.*;

import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.FloatBuffer;
import java.util.Collections;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * Runs inference on the DocQuadNet-256 ONNX model using ONNX Runtime.
 * <p>
 * Input:  {@code float[3 * 256 * 256]} in NCHW layout, RGB, values 0..1<br>
 * Output: {@code corner_heatmaps [1,4,64,64]} + {@code mask_logits [1,1,64,64]}
 * <p>
 * Thread-safe singleton access via {@link #getInstance(Context, String)}.
 * <p>
 * <b>Android dependency:</b> Uses {@code Context} for asset loading and {@code Log} for logging.
 * Adapt these two points when porting to a non-Android environment.
 */
public final class DocQuadOrtRunner implements AutoCloseable {

    private static final String TAG = "DocQuadOrtRunner";

    public static final int IN_H = 256;
    public static final int IN_W = 256;
    public static final int OUT_H = 64;
    public static final int OUT_W = 64;

    private static volatile DocQuadOrtRunner instance;
    private static final Object LOCK = new Object();
    private static final Executor DEFAULT_EXECUTOR = Executors.newSingleThreadExecutor();

    private final OrtEnvironment env;
    private final OrtSession session;

    private DocQuadOrtRunner(Context context, String modelAssetPath) throws Exception {
        this.env = OrtEnvironment.getEnvironment();
        File modelFile = copyAssetToCache(context, modelAssetPath);

        try (OrtSession.SessionOptions opts = new OrtSession.SessionOptions()) {
            opts.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT);
            opts.setIntraOpNumThreads(Math.max(1, Runtime.getRuntime().availableProcessors() / 2));
            try { opts.addNnapi(); Log.i(TAG, "NNAPI EP enabled"); }
            catch (Throwable t) { Log.i(TAG, "NNAPI not available: " + t.getMessage()); }
            try { opts.addXnnpack(Collections.emptyMap()); Log.i(TAG, "XNNPACK EP enabled"); }
            catch (Throwable t) { Log.i(TAG, "XNNPACK not available: " + t.getMessage()); }
            this.session = env.createSession(modelFile.getAbsolutePath(), opts);
        }
        Log.d(TAG, "Model loaded from " + modelFile.getAbsolutePath());
    }

    /** Returns the singleton instance (thread-safe, lazy init). */
    public static DocQuadOrtRunner getInstance(Context context, String modelAssetPath) throws Exception {
        if (instance == null) {
            synchronized (LOCK) {
                if (instance == null) {
                    instance = new DocQuadOrtRunner(context, modelAssetPath);
                }
            }
        }
        return instance;
    }

    /** Async variant using default single-thread executor. */
    public static CompletableFuture<DocQuadOrtRunner> getInstanceAsync(Context context, String modelAssetPath) {
        return getInstanceAsync(context, modelAssetPath, DEFAULT_EXECUTOR);
    }

    /** Async variant with custom executor. */
    public static CompletableFuture<DocQuadOrtRunner> getInstanceAsync(Context context, String modelAssetPath, Executor executor) {
        return CompletableFuture.supplyAsync(() -> {
            try { return getInstance(context, modelAssetPath); }
            catch (Exception e) { throw new java.util.concurrent.CompletionException(e); }
        }, executor);
    }

    /** Returns true if the singleton is already loaded. */
    public static boolean isInstanceLoaded() {
        return instance != null;
    }

    /** Releases the singleton and closes the ONNX session. */
    public static void releaseInstance() {
        synchronized (LOCK) {
            if (instance != null) {
                try { instance.close(); } catch (Exception e) { Log.w(TAG, "Error closing: " + e.getMessage()); }
                instance = null;
            }
        }
    }

    /**
     * Runs inference.
     *
     * @param inputNchw float array of length {@code 3 * 256 * 256}, NCHW, RGB, 0..1
     * @return model outputs (corner heatmaps + mask logits)
     */
    public Outputs run(float[] inputNchw) throws Exception {
        if (inputNchw == null || inputNchw.length != 3 * IN_H * IN_W)
            throw new IllegalArgumentException("inputNchw must have length " + (3 * IN_H * IN_W));

        long[] inputShape = {1, 3, IN_H, IN_W};
        try (OnnxTensor input = OnnxTensor.createTensor(env, FloatBuffer.wrap(inputNchw), inputShape);
             OrtSession.Result results = session.run(Collections.singletonMap("input", input))) {

            float[][][][] maskLogits = getRequiredFloat4d(results, "mask_logits");
            float[][][][] cornerHeatmaps = getRequiredFloat4d(results, "corner_heatmaps");
            return new Outputs(maskLogits, cornerHeatmaps);
        }
    }

    @Override
    public void close() throws Exception {
        if (session != null) session.close();
    }

    // ── Asset handling ──

    private static File copyAssetToCache(Context context, String assetPath) throws IOException {
        AssetManager am = context.getAssets();
        String baseName = new File(assetPath).getName();
        long versionCode;
        try {
            android.content.pm.PackageInfo pi =
                    context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
            versionCode = pi.getLongVersionCode();
        } catch (Exception e) { versionCode = -1L; }

        String versionedName = versionCode + "_" + baseName;
        File outFile = new File(context.getCacheDir(), versionedName);
        if (!outFile.exists()) {
            Log.i(TAG, "Copying asset " + assetPath + " to cache as " + versionedName);
            try (InputStream is = am.open(assetPath);
                 FileOutputStream fos = new FileOutputStream(outFile)) {
                byte[] buffer = new byte[256 * 1024];
                int len;
                while ((len = is.read(buffer)) != -1) fos.write(buffer, 0, len);
            }
            deleteStaleModelFiles(context.getCacheDir(), baseName, versionedName);
        }
        return outFile;
    }

    private static void deleteStaleModelFiles(File cacheDir, String baseName, String currentName) {
        File[] staleFiles = cacheDir.listFiles((dir, name) ->
                name.endsWith("_" + baseName) && !name.equals(currentName));
        if (staleFiles != null) {
            for (File stale : staleFiles) {
                if (stale.delete()) Log.i(TAG, "Deleted stale: " + stale.getName());
                else Log.w(TAG, "Failed to delete stale: " + stale.getName());
            }
        }
    }

    // ── Output extraction ──

    private static float[][][][] getRequiredFloat4d(OrtSession.Result results, String outputName) throws OrtException {
        Optional<OnnxValue> ov = results.get(outputName);
        if (ov.isEmpty()) throw new IllegalStateException("Missing output: " + outputName);
        OnnxValue val = ov.get();
        if (!(val instanceof OnnxTensor)) throw new IllegalStateException(outputName + " is not a tensor");
        Object raw = ((OnnxTensor) val).getValue();
        if (!(raw instanceof float[][][][]))
            throw new IllegalStateException(outputName + " is not float[][][][]");
        return (float[][][][]) raw;
    }

    /** Model outputs container. */
    public static final class Outputs {
        private final float[][][][] maskLogits;
        private final float[][][][] cornerHeatmaps;

        public Outputs(float[][][][] maskLogits, float[][][][] cornerHeatmaps) {
            this.maskLogits = maskLogits;
            this.cornerHeatmaps = cornerHeatmaps;
        }

        public float[][][][] maskLogits() { return maskLogits; }
        public float[][][][] cornerHeatmaps() { return cornerHeatmaps; }
    }
}

File: plugin-nativeprocessor/platforms/android/java/com/akylas/documentscanner/docquad/DocQuadOrtRunner.kt
package com.akylas.documentscanner.docquad

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OnnxValue
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession

import android.content.Context
import android.util.Log

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.FloatBuffer
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * Runs inference on the DocQuadNet-256 ONNX model using ONNX Runtime.
 *
 * Input:  `float[3 * 256 * 256]` in NCHW layout, RGB, values 0..1
 * Output: `corner_heatmaps [1,4,64,64]` + `mask_logits [1,1,64,64]`
 *
 * Thread-safe singleton access via [getInstance].
 *
 * **Android dependency:** Uses `Context` for asset loading and `Log` for logging.
 */
class DocQuadOrtRunner private constructor(context: Context, modelAssetPath: String) : AutoCloseable {

    private val env: OrtEnvironment = OrtEnvironment.getEnvironment()
    private val session: OrtSession

    init {
        val modelFile = copyAssetToCache(context, modelAssetPath)

        val opts = OrtSession.SessionOptions()
        try {
            opts.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
            opts.setIntraOpNumThreads(maxOf(1, Runtime.getRuntime().availableProcessors() / 2))
            try { opts.addNnapi(); Log.i(TAG, "NNAPI EP enabled") }
            catch (t: Throwable) { Log.i(TAG, "NNAPI not available: ${t.message}") }
            try { opts.addXnnpack(emptyMap()); Log.i(TAG, "XNNPACK EP enabled") }
            catch (t: Throwable) { Log.i(TAG, "XNNPACK not available: ${t.message}") }
            session = env.createSession(modelFile.absolutePath, opts)
        } finally {
            opts.close()
        }
        Log.d(TAG, "Model loaded from ${modelFile.absolutePath}")
    }

    /**
     * Runs inference.
     *
     * @param inputNchw float array of length `3 * 256 * 256`, NCHW, RGB, 0..1
     * @return model outputs (corner heatmaps + mask logits)
     */
    fun run(inputNchw: FloatArray): Outputs {
        require(inputNchw.size == 3 * IN_H * IN_W) { "inputNchw must have length ${3 * IN_H * IN_W}" }

        val inputShape = longArrayOf(1, 3, IN_H.toLong(), IN_W.toLong())
        OnnxTensor.createTensor(env, FloatBuffer.wrap(inputNchw), inputShape).use { input ->
            session.run(mapOf("input" to input)).use { results ->
                val maskLogits = getRequiredFloat4d(results, "mask_logits")
                val cornerHeatmaps = getRequiredFloat4d(results, "corner_heatmaps")
                return Outputs(maskLogits, cornerHeatmaps)
            }
        }
    }

    override fun close() {
        session.close()
    }

    // ── Asset handling ──

    companion object {
        private const val TAG = "DocQuadOrtRunner"

        const val IN_H = 256
        const val IN_W = 256
        const val OUT_H = 64
        const val OUT_W = 64

        @Volatile
        private var instance: DocQuadOrtRunner? = null
        private val LOCK = Any()
        private val DEFAULT_EXECUTOR: Executor = Executors.newSingleThreadExecutor()

        /** Returns the singleton instance (thread-safe, lazy init). */
        @JvmStatic
        @Throws(Exception::class)
        fun getInstance(context: Context, modelAssetPath: String): DocQuadOrtRunner {
            if (instance == null) {
                synchronized(LOCK) {
                    if (instance == null) {
                        instance = DocQuadOrtRunner(context, modelAssetPath)
                    }
                }
            }
            return instance!!
        }

        /** Async variant using default single-thread executor. */
        @JvmStatic
        fun getInstanceAsync(context: Context, modelAssetPath: String): CompletableFuture<DocQuadOrtRunner> {
            return getInstanceAsync(context, modelAssetPath, DEFAULT_EXECUTOR)
        }

        /** Async variant with custom executor. */
        @JvmStatic
        fun getInstanceAsync(context: Context, modelAssetPath: String, executor: Executor): CompletableFuture<DocQuadOrtRunner> {
            return CompletableFuture.supplyAsync({
                try { getInstance(context, modelAssetPath) }
                catch (e: Exception) { throw java.util.concurrent.CompletionException(e) }
            }, executor)
        }

        /** Returns true if the singleton is already loaded. */
        @JvmStatic
        fun isInstanceLoaded(): Boolean = instance != null

        /** Releases the singleton and closes the ONNX session. */
        @JvmStatic
        fun releaseInstance() {
            synchronized(LOCK) {
                instance?.let {
                    try { it.close() } catch (e: Exception) { Log.w(TAG, "Error closing: ${e.message}") }
                    instance = null
                }
            }
        }

        @Throws(IOException::class)
        private fun copyAssetToCache(context: Context, assetPath: String): File {
            val am = context.assets
            val baseName = File(assetPath).name
            val versionCode: Long = try {
                val pi = context.packageManager.getPackageInfo(context.packageName, 0)
                pi.longVersionCode
            } catch (e: Exception) { -1L }

            val versionedName = "${versionCode}_$baseName"
            val outFile = File(context.cacheDir, versionedName)
            if (!outFile.exists()) {
                Log.i(TAG, "Copying asset $assetPath to cache as $versionedName")
                am.open(assetPath).use { inputStream ->
                    FileOutputStream(outFile).use { fos ->
                        val buffer = ByteArray(256 * 1024)
                        var len: Int
                        while (inputStream.read(buffer).also { len = it } != -1) {
                            fos.write(buffer, 0, len)
                        }
                    }
                }
                deleteStaleModelFiles(context.cacheDir, baseName, versionedName)
            }
            return outFile
        }

        private fun deleteStaleModelFiles(cacheDir: File, baseName: String, currentName: String) {
            val staleFiles = cacheDir.listFiles { _, name ->
                name.endsWith("_$baseName") && name != currentName
            }
            staleFiles?.forEach { stale ->
                if (stale.delete()) Log.i(TAG, "Deleted stale: ${stale.name}")
                else Log.w(TAG, "Failed to delete stale: ${stale.name}")
            }
        }

        private fun getRequiredFloat4d(results: OrtSession.Result, outputName: String): Array<Array<Array<FloatArray>>> {
            val ov = results.get(outputName)
            check(ov.isPresent) { "Missing output: $outputName" }
            val value = ov.get()
            check(value is OnnxTensor) { "$outputName is not a tensor" }
            val raw = value.value
            check(raw is Array<*>) { "$outputName is not float[][][][]" }
            @Suppress("UNCHECKED_CAST")
            return raw as Array<Array<Array<FloatArray>>>
        }
    }

    /** Model outputs container. */
    data class Outputs(
        val maskLogits: Array<Array<Array<FloatArray>>>,
        val cornerHeatmaps: Array<Array<Array<FloatArray>>>
    )
}
